crates/adapters/src/secret_resolver.rs

```diff
@@ -0,0 +1,387 @@
+                .collect::<Result<Vec<Value>, SecretRefResolutionError>>()?,
+        ),
+        Value::Mapping(mapping) => {
+            let mut new_mapping = Mapping::new();
+            for (k, v) in mapping.into_iter() {
+                if let Some(_existing) =
+                    new_mapping.insert(k, resolve_secret_references_in_yaml(secrets_dir, v)?)
+                {
+                    return Err(SecretRefResolutionError::DuplicateKeyInMapping);
+                }
+            }
+            Value::Mapping(new_mapping)
+        }
+        Value::Tagged(mut tag_val) => {
+            tag_val.value = resolve_secret_references_in_yaml(secrets_dir, tag_val.value)?;
+            Value::Tagged(tag_val)
+        }
+    })
+}
+
+/// Resolves a string which can potentially be a secret reference.
+fn resolve_potential_secret_reference_string(
+    secrets_dir: &Path,
+    s: String,
+) -> Result<String, SecretRefResolutionError> {
+    match MaybeSecretRef::new(s) {
+        Ok(maybe_secret_ref) => match maybe_secret_ref {
+            MaybeSecretRef::String(plain_str) => Ok(plain_str),
+            MaybeSecretRef::SecretRef(secret_ref) => match &secret_ref {
+                SecretRef::Kubernetes { name, data_key } => {
+                    // Secret reference: `${secret:kubernetes:<name>/<data key>}`
+                    // File location: `<secrets dir>/kubernetes/<name>/<data key>`
+                    let path = Path::new(secrets_dir)
+                        .join("kubernetes")
+                        .join(name)
+                        .join(data_key);
+
+                    // If the file does not exist or is not a file, produce a custom error informing
+                    // of this fact rather than generically stating we are unable to read.
+                    if path.is_file() {
+                        match fs::read_to_string(&path) {
+                            Ok(content) => Ok(content),
+                            Err(_e) => {
+                                // The read error is not part of the error returned to prevent
+                                // displaying any of the secret content
+                                Err(SecretRefResolutionError::CannotReadSecretFile {
+                                    secret_ref,
+                                    path: path.display().to_string(),
+                                })
+                            }
+                        }
+                    } else if path.exists() {
+                        Err(SecretRefResolutionError::SecretPathIsNotFile {
+                            secret_ref,
+                            path: path.display().to_string(),
+                        })
+                    } else {
+                        Err(SecretRefResolutionError::SecretFileDoesNotExist {
+                            secret_ref,
+                            path: path.display().to_string(),
+                        })
```

Comment:
Because of how exists works, this will also report the secret file does not exist if there's a permission error, broken links etc. Maybe that's OK but I'd consider try_exists instead to report the error more accurately.